### PR TITLE
Adding summary metric for scheduling latency

### PR DESCRIPTION
**Pull request by krzysied**
**What this PR does / why we need it**:
Re-introduces histogram metrics for the backward compatibility.
Changes SchedulingLatency metric to satisfy prometheus best practice.
ref #64316

```release-note
Add a summary-type metric 'scheduling_latency_seconds' in scheduler.
```

### CODE REVIEW

**Comment by shyamjvs** [comment]
That lgtm.

/cc @bsalamat @misterikkit 
Could you PTAL at the change? This is re-adding the old metrics which were earlier removed in https://github.com/kubernetes/kubernetes/pull/64316.

**Comment by shyamjvs** [comment]
/sig scheduling
/kind bug
/priority important-soon
/milestone v1.11
/status approved-for-milestone

As this is an important fix to go in (in order to avoid backward incompatible change in the release).

File: pkg/scheduler/metrics/metrics.go [hunk 1]
```diff
@@ -27,7 +27,7 @@ const (
 	// SchedulerSubsystem - subsystem name used by scheduler
 	SchedulerSubsystem = "scheduler"
 	// SchedulingLatencyName - scheduler latency metric name
-	SchedulingLatencyName = "scheduling_latencies_summary"
+	SchedulingLatencyName = "scheduling_latency_seconds"
 
 	// OperationLabel - operation label name
 	OperationLabel = "operation"
```
Comment (misterikkit) [inline pkg/scheduler/metrics/metrics.go]: In the [naming](https://prometheus.io/docs/practices/naming/) section of the prometheus guide, they mention that,

> either the sum() or the avg() over all dimensions of a given metric should be meaningful.

When you define a label with values for "binding", "algorithm", and "e2e", then sum and avg are both meaningless.

For counters, it's clear to see how the "total" would be represented by the sum over each label value. I am not sure if histograms work the same way, but this smells odd to me.

Comment (misterikkit) [inline pkg/scheduler/metrics/metrics.go]: I would prefer that we don't use `scheduling_algorithm` as a bucket name. "scheduling" adds no meaning in this context, as the subsystem name and metric name already contain it. "algorithm" is vague to anyone not familiar with our internal implementation details.

Maybe `select_node` instead?

**Review by misterikkit — COMMENTED** [review]
I would prefer re-adding the old metrics and introducing new metrics to be at least in separate git commits, if not separate PRs. If something needs to be reverted for some reason, I don't want them to have a shared fate.

**Comment by misterikkit** [comment]
Side note: I left some code review comments on lines that were not changed in this PR, but that are part of the new metrics. The review would have been easier if we rolled back the original PR (restoring old names) and then re-uploaded the new metrics PR.

File: pkg/scheduler/metrics/metrics.go (hunk 1 above)
Comment (krzysied) [reply pkg/scheduler/metrics/metrics.go]: Renamed to "selecting_node"

Comment (krzysied) [reply pkg/scheduler/metrics/metrics.go]: Ok. I think I understand your point.
I've removed the "e2e" label. This "e2e" is equivalent to the sum of "binding" and "algorithm".

File: pkg/scheduler/metrics/metrics.go [hunk 2]
```diff
@@ -118,3 +145,8 @@ func Reset() {
 func SinceInMicroseconds(start time.Time) float64 {
 	return float64(time.Since(start).Nanoseconds() / time.Microsecond.Nanoseconds())
 }
+
+// SinceInSeconds gets the time since the specified start in seconds.
+func SinceInSeconds(start time.Time) float64 {
+	return float64(time.Since(start).Nanoseconds()) / float64(time.Second.Nanoseconds())
```
[the remark this answers is not in the document]
Comment (krzysied) [reply pkg/scheduler/metrics/metrics.go]: done

File: pkg/scheduler/metrics/metrics.go [hunk 3]
```diff
@@ -41,6 +41,30 @@ const (
...
+		prometheus.HistogramOpts{
+			Subsystem: SchedulerSubsystem,
+			Name:      "binding_latency_microseconds",
+			Help:      "Binding latency",
+			Buckets:   prometheus.ExponentialBuckets(1000, 2, 15),
+		},
+	)
 	SchedulingLatency = prometheus.NewSummaryVec(
 		prometheus.SummaryOpts{
 			Subsystem: SchedulerSubsystem,
```
[the remark this answers is not in the document]
Comment (krzysied) [reply pkg/scheduler/metrics/metrics.go]: fixed

File: test/e2e/framework/metrics_util.go [hunk 4]
```diff
@@ -525,7 +525,7 @@ func getSchedulingLatency(c clientset.Interface) (*SchedulingMetrics, error) {
 		if err != nil {
 			return nil, err
 		}
-		setQuantile(metric, quantile, time.Duration(int64(latency)))
+		setQuantile(metric, quantile, time.Duration(int64(latency*1e6)))
```
[the remark this answers is not in the document]
Comment (krzysied) [reply test/e2e/framework/metrics_util.go]: done

**Comment by krzysied** [comment]
@misterikkit Now we have two separated commits. One for reverting changes from previous PR, other one for re-introducing summary metrics.

File: pkg/scheduler/metrics/metrics.go [hunk 5]
```diff
@@ -23,70 +23,105 @@ import (
...
+	SchedulerSubsystem = "scheduler"
+	// SchedulingLatencyName - scheduler latency metric name
+	SchedulingLatencyName = "scheduling_latency_seconds"
+
+	// OperationLabel - operation label name
+	OperationLabel = "operation"
+	// Binding - binding operation label value
+	Binding = "binding"
+	// SchedulingAlgorithm - scheduling algorithm operation label value
+	SchedulingAlgorithm = "selecting_node"
```
Comment (misterikkit) [inline pkg/scheduler/metrics/metrics.go]: Please rename the variable to match the string.

File: pkg/scheduler/metrics/metrics.go [hunk 6]
```diff
@@ -95,19 +130,23 @@ var registerMetrics sync.Once
...
-		prometheus.MustRegister(PreemptionAttempts)
+		for _, metric := range metricsList {
+			prometheus.MustRegister(metric)
+		}
 	})
 }
 
+// Reset resets metrics
+func Reset() {
+	SchedulingLatency.Reset()
```
Comment (misterikkit) [inline pkg/scheduler/metrics/metrics.go]: Should this reset everything? (Is this a normal prometheus pattern?)

**Comment by misterikkit** [comment]
> bumping priority because code freeze

Sorry, the review is taking some time because we are fixing existing stuff and adding new stuff in the same PR.

File: pkg/scheduler/metrics/metrics.go (hunk 6 above)
Comment (krzysied) [reply pkg/scheduler/metrics/metrics.go]: The problem is that simple metrics don't have reset function, only vectors do.

File: pkg/scheduler/metrics/metrics.go (hunk 5 above)
Comment (krzysied) [reply pkg/scheduler/metrics/metrics.go]: done

**Comment by shyamjvs** [comment]
@krzysied Could you rename the PR title to be more apt (after the newer changes)? Sth like 'Adding summary metric for scheduling latency'.

**Comment by dims** [comment]
@krzysied @shyamjvs this has bit-rotted, can you please fix it up? (and line up lgtm/approvals too)

**Comment by shyamjvs** [comment]
@dims We should definitely get this in for the release.

@misterikkit Could you review this PR soonish?
Also @krzysied - seems like you'd need to rebase (there are some conflicts).

**Comment by krzysied** [comment]
@shyamjvs I've done the rebase. There should be no conflicts now.

File: pkg/scheduler/metrics/metrics.go [hunk 7]
```diff
@@ -76,6 +91,14 @@ var (
 			Buckets:   prometheus.ExponentialBuckets(1000, 2, 15),
 		},
 	)
+	BindingLatency = prometheus.NewHistogram(
```
Comment (bsalamat) [inline pkg/scheduler/metrics/metrics.go]: Looks like this metric is now deprecated and replaced by `Binding`. Could you please add a comment to mark it deprecated?

**Review by bsalamat — COMMENTED** [review]
Looks good. Just a minor comment.

**Comment by jberkus** [comment]
Folks:

This needs to get approved and merged by tommorrow, or we should punt it to the next milestone.  Thanks!

File: pkg/scheduler/metrics/metrics.go (hunk 7 above)
Comment (krzysied) [reply pkg/scheduler/metrics/metrics.go]: Imo, this metric shouldn't be deprecated - histogram metric provides different data than summary metric. I think both metrics are useful and could be used to measure different aspects.

**Comment by shyamjvs** [comment]
Since we seem to be bound by time, let's get this in - it's fixing a backward-incompatible change.

**Comment by shyamjvs** [comment]
FYI - I've modified the PR description to add a release-note for this change (as I think it needs one).

**Pull request merged**